Add integration tests that run against a real Redshift cluster

JoshRosen opened the pull request:
This PR adds end-to-end integration tests that run against a real Redshift cluster.

I moved the build definition from `.sbt` to `.scala` format in order to use SBT's standard [IntegrationTests](http://www.scala-sbt.org/0.13.0/docs/Detailed-Topics/Testing#integration-tests) configuration.

I configured the Travis build to store Redshift credentials using [encrypted environment variables](http://docs.travis-ci.com/user/environment-variables/#Encrypted-Variables). To minimize security risks, Travis does not allow pull requests from third-party forks to run tests using these credentials.

JoshRosen commented:
Alright, the basic test is now passing (fixed the non-deterministic failure issue), so I'm now going to focus on enabling more tests.

JoshRosen commented on .travis.yml:
```diff
@@ -8,10 +8,25 @@ cache:
+    - secure: "LIkY/ZpBXK3vSFsdpBSRXEsgfD2wDF52X8OZOlyBJOiZpS4y1/obj8b3VQABDPyPH95bGX/LOpM0vVM137rYgF0pskgVEzLMyZOPpwYqNGPf/d4BtQhBRc8f7+jmr6D4Hrox4jCl0cCKaeiTazun2+Y9E+zgCUDvQ8y9qGctR2k="
+    # AWS_ACCESS_KEY_ID
+    - secure: "CDlql+nrgdi7sUr7bYyXF4CFoOUCiJG9WEYNRV4k/lC37eS/al3iVYicnXqF+6UrPv5a4kHulG4d3g78J4hzn4ZVJuEhn6v8beoOBUoJJ7W/J05hVwGiQFxUq86wT3tIaBrAuDmOXaAnPEvDmPfJGNZL9ZG1CaQJo70R/HkbbVA="
+    # AWS_SECRET_ACCESS_KEY
```
I wonder if I should purposely give these different names than the ones expected by the Amazon SDKs; this might be necessary in order to be able to test credential mechanisms (e.g. for writing a regression test for #32)

JoshRosen commented:
All tests are now enabled, but some seem to fail due to problems in the tests themselves (due to not using a HiveContext) or due to bugs.

JoshRosen commented:
This is now ready for review.  Let's not commit until #42 goes in and until I have time for one final pass. This isn't all of the tests that I want to add but it's a reasonable first cut; I'll add more tests and fix more potential bugs in smaller followups.

marmbrus commented on .travis.yml:
```diff
@@ -8,12 +8,28 @@ cache:
+    # AWS_S3_SCRATCH_SPACE
+    - secure: "LvndQIW6dHs6nyaMHtblGI/oL+s460lOezFs2BoD0Isenb/O/IM+nY5K9HepTXjJIcq8qvUYnojZX1FCrxxOXX2/+/Iihiq7GzJYdmdMC6hLg9bJYeAFk0dWYT88/AwadrJCBOa3ockRLhiO3dkai7Ki5+M1erfaFiAHHMpJxYQ="
 script:
   - sbt -Dhadoop.version=$HADOOP_VERSION coverage test
```
Nit: is there a reason to not test style and other fast stuff first?

JoshRosen replied:
I wanted the tests to still be able to run even if style checks failed. I figured this wasn't a huge deal here compared to what we do in Spark because the test time is really fast.

marmbrus commented on project/SparkRedshiftBuild.scala:
```diff
@@ -0,0 +1,68 @@
+      name := "spark-redshift",
+      organization := "com.databricks",
+      version := "0.4.1-SNAPSHOT",
+      scalaVersion := "2.10.4",
```
Doesn't have to be here, but in a followup can we make sure to cross publish 2.11?

JoshRosen replied:
Yep, I plan to add Scala versions to the Matrix build configuration.

marmbrus commented on project/SparkRedshiftBuild.scala:
```diff
@@ -0,0 +1,68 @@
+      libraryDependencies ++= Seq(
+        "com.amazonaws" % "aws-java-sdk-core" % "1.9.40" % "provided",
+        // We require spark-avro, but avro-mapred must be provided to match Hadoop version:
+        "com.databricks" %% "spark-avro" % "1.0.0",
```
Another comment so I don't forget, we should update this when we publish a new version of avro (and I guess this creates a cross publishing ordering dependency).

JoshRosen replied:
Yes.  The fact that `spark-avro` is not currently cross-published for 2.11 is one of the main blockers to 2.11 support here.

marmbrus commented on src/it/scala/com/databricks/spark/redshift/RedshiftIntegrationSuite.scala:
```diff
@@ -0,0 +1,335 @@
+    val df = sqlContext.createDataFrame(rdd, TestUtils.testSchema)
+
+    // Check that SaveMode.ErrorIfExists throws an exception
+    intercept[Exception] {
```
`AnalysisException`?

JoshRosen replied:
Carryover from the other tests that these were inspired by.  I'll fix in both suites.

marmbrus commented on src/it/scala/com/databricks/spark/redshift/RedshiftIntegrationSuite.scala:
```diff
@@ -0,0 +1,335 @@
+      sqlContext.sql("select * from test_table order by testbyte, testbool"),
+      TestUtils.expectedData)
+  }
+}
```
nit: newline

marmbrus commented:
This is awesome! LGTM.

JoshRosen commented:
This was rebased on top of #32, but I'm going to try to back out of that rebase so that we can write a proper regression test for that issue.

JoshRosen commented:
Alright, I have now backed out all non-test-related changes and have addressed the review comments, so I'm going to merge this now and will continue to iterate in followups.
